### ISSUE TITLE
[java] Non-static initializers and methods should discourage use of "this" as parameter.

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2030]
**Affects PMD Version:** 
All.

**Rule:**
Non-static methods and constructors should not use "this" as a parameter.

**Description:**
Best described here:
https://twitter.com/heinzkabutz/status/117528379359223398

**Comment by oowekyala** [comment]
Your link is dead, but I found the thread. The receiver parameter is not necessarily useless, as Brian Goetz himself argues in that thread. It's only useless if it is not annotated (or just annotated `NonNull` or a variant).

Feel free to submit a PR

**Comment by Fernal73** [comment]
Yep, I updated the documentation. If submitting a PR involves coding a solution, Nope.

**Comment by Fernal73** [comment]
Added code sample to description.

**Comment by Fernal73** [comment]
https://stackoverflow.com/questions/49533897/is-a-method-with-receiver-parameter-override-equivalent-to-the-same-method-decla

**Comment by Fernal73** [comment]
https://github.com/pmd/pmd/issues/719

**Comment by Fernal73** [comment]
https://bugs.openjdk.java.net/browse/JDK-8195646

**Comment by Fernal73** [comment]
http://mail.openjdk.java.net/pipermail/core-libs-dev/2019-June/060767.html

**Comment by Fernal73** [comment]
https://coderanch.com/t/708483/java/understanding-implicit-parameter

**Comment by Fernal73** [comment]
https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8175198

**Comment by Fernal73** [comment]
https://github.com/checkstyle/checkstyle/issues/3239

**Comment by Fernal73** [comment]
https://stackoverflow.com/questions/49533897/is-a-method-with-receiver-parameter-override-equivalent-to-the-same-method-decla

**Comment by Fernal73** [comment]
https://github.com/Fernal73/LearnJava/blob/master/Reflection/reflection/SyntheticSample.java

A synthetic method in an inner class displays it's signature as consisting of a call to a receiver parameter.

**Comment by adangel** [comment]
FYI - I've updated to description to have two examples: one the new rule should detect (e.g. "invalid code") and another one, that should produce no violation (e.g. "valid code")

**Comment by Fernal73** [comment]
Nice. The link that accompanies the valid code example is pretty useful as
well.

**Comment by Fernal73** [comment]
Is this going to be in PMD 7?

**Comment by UncleOwen** [comment]
Why should Foo.bar2() be reported in the first example? It's a receiver parameter, and it's annotated.

**Comment by adangel** [comment]
> Why should Foo.bar2() be reported in the first example? It's a receiver parameter, and it's annotated.

In that case, the receiver parameter is annotated with NonNull - which is unnecessary, as the receiver parameter will never be null (`this` can't be null, it always exists).

**Comment by UncleOwen** [comment]
Thanks. I've added that information to the OP.
